mmd_tools/core/vmd/importer.py

```diff
@@ -291,6 +292,57 @@ def __fixFcurveHandles(fcurve):
         kp.handle_right_type = 'FREE'
         kp.handle_right = kp.co + Vector((1, 0))
 
+    @staticmethod
+    def __convertQuaternionFCurveToEulerFCurve(action, euler_bones):
+        remove_fcurves = []
+        for bone in euler_bones:
+            axis_order = bone.rotation_mode
+            quaternion_data_path = 'pose.bones["%s"].rotation_quaternion' % bone.name
+            euler_data_path = 'pose.bones["%s"].rotation_euler' % bone.name
+
+            quaternion_fcurves = []
+            euler_fcurves = []
+            for axis_i in range(4):
+                c = action.fcurves.find(data_path=quaternion_data_path, index=axis_i)
+                if c is not None:
+                    quaternion_fcurves.append(c)
+            if len(quaternion_fcurves) != 4:
+                continue # no keyframes
+            for axis_i in range(3):
+                euler_fcurves.append(action.fcurves.new(data_path=euler_data_path, index=axis_i, action_group=bone.name))
+
+            keyframe_points = defaultdict(lambda: [None, None, None, None])
+            for c in quaternion_fcurves:
+                for p in c.keyframe_points:
+                    keyframe_points[int(p.co[0])][c.array_index] = p
+            for frame, p in keyframe_points.items():
+                rot = Quaternion((p[0].co[1], p[1].co[1], p[2].co[1], p[3].co[1])).to_euler(axis_order)
+                for c, value in zip(euler_fcurves, rot):
+                    c.keyframe_points.insert(frame, value)
+
+            # FIXME: interpolation
+            for qc, ec in zip(quaternion_fcurves[1:], euler_fcurves):
+                qv = [p.co[1] for p in qc.keyframe_points]
+                ev = [p.co[1] for p in ec.keyframe_points]
+                q_range = max(qv) - min(qv)
+                e_range = max(ev) - min(ev)
+                if q_range != 0:
+                    scale = e_range / q_range
+                else:
+                    scale = 1
+                for qp, ep in zip(qc.keyframe_points, ec.keyframe_points):
+                    ep.interpolation = qp.interpolation
+                    ep.handle_right_type = qp.handle_right_type
+                    ep.handle_left_type = qp.handle_left_type
+                    ep.handle_right = Vector((qp.handle_right[0], (qp.handle_right[1] - qp.co[1]) * scale + ep.co[1]))
+                    ep.handle_left = Vector((qp.handle_left[0], (qp.handle_left[1] - qp.co[1]) * scale + ep.co[1]))
```

Comment:
It can be fixed by the following code. :smile:
```python
            def _get_bezier_data(prev_points, curr_points):
                for qp_prev, qp in zip(prev_points, curr_points):
                    delta_y = qp.co[1] - qp_prev.co[1]
                    if abs(delta_y) > 1e-4:
                        return qp_prev.interpolation, ((qp_prev.handle_right[1]-qp_prev.co[1]) / delta_y), ((qp.handle_left[1]-qp_prev.co[1]) / delta_y)
                return 'LINEAR', 0, 1

            prev_points = ()
            last_rot = bone.rotation_euler
            key_func = lambda kp: kp.co[0]
            for qw, qx, qy, qz in zip(*[sorted(c.keyframe_points, key=key_func) for c in quaternion_fcurves]):
                rot = Quaternion((qw.co[1], qx.co[1], qy.co[1], qz.co[1])).to_euler(axis_order)
                rot.make_compatible(last_rot)
                frame = qw.co[0]
                curr_points = (qx, qy, qz, qw)
                if prev_points:
                    interpolation, y1, y2 = _get_bezier_data(prev_points, curr_points)
                    for c, value in zip(euler_fcurves, rot):
                        ep = c.keyframe_points.insert(frame, value)
                        ep.handle_left.x = qw.handle_left.x
                        ep.handle_right.x = qw.handle_right.x
                        ep_prev = c.keyframe_points[-2]
                        y_base = ep_prev.co[1]
                        scale = value - y_base
                        ep_prev.interpolation = interpolation
                        ep_prev.handle_right_type = 'FREE'
                        ep_prev.handle_right.y = y1 * scale + y_base
                        ep.handle_left_type = 'FREE'
                        ep.handle_left.y = y2 * scale + y_base
                else:
                    for c, value in zip(euler_fcurves, rot):
                        ep = c.keyframe_points.insert(frame, value)
                        ep.handle_left.x = qw.handle_left.x
                        ep.handle_right.x = qw.handle_right.x
                prev_points = curr_points
                last_rot = rot
```

Reply:
I'm checking your code and encountering a strange problem. 
The problem is that the frame of the handle (*_handle.x) is different from the original quaternion fcurve, but I am not sure where it is replaced/broken.

I added debug print code.
```python
                if prev_points:                                                                           
                    print("***")                                                                          
                    interpolation, y1, y2 = _get_bezier_data(prev_points, curr_points)                    
                    for c, value in zip(euler_fcurves, rot):                                              
                        ep = c.keyframe_points.insert(frame, value)                                       
                        ep.handle_left.x = qw.handle_left.x                                               
                        ep.handle_right.x = qw.handle_right.x                                             
                        ep_prev = c.keyframe_points[-2]                                                   
                        y_base = ep_prev.co[1]                                                            
                        scale = value - y_base                                                            
                        ep_prev.interpolation = interpolation                                             
                        ep_prev.handle_right_type = 'FREE'                                                
                        ep_prev.handle_right.y = y1 * scale + y_base                                      
                        ep.handle_left_type = 'FREE'                                                      
                        ep.handle_left.y = y2 * scale + y_base                                            
                        print("handle_right x", "cur", ep.handle_right.x, "prev", ep_prev.handle_right.x) 

```
output
```
***
handle_right x cur 33.031494140625 prev 9.333333015441895
handle_right x cur 33.031494140625 prev 9.333333015441895
handle_right x cur 33.031494140625 prev 9.333333015441895
***
handle_right x cur 90.6377944946289 prev 41.66666793823242
handle_right x cur 90.6377944946289 prev 41.66666793823242
handle_right x cur 90.6377944946289 prev 41.66666793823242
***
handle_right x cur 185.99212646484375 prev 110.33332824707031
handle_right x cur 185.99212646484375 prev 110.33332824707031
handle_right x cur 185.99212646484375 prev 110.33332824707031
....
```
The second keyframe's `33.031494140625` is correct `handle_right.x` and is set correctly, but the next time it is called, it will be replaced by `prev 41.66666793823242`.

So I changed to add the keyframes to fcurve serially.
```python
            def _get_bezier_data(prev_points, curr_points):
                for qp_prev, qp in zip(prev_points, curr_points):
                    delta_y = qp.co[1] - qp_prev.co[1]
                    if abs(delta_y) > 1e-4:
                        return qp_prev.interpolation, ((qp_prev.handle_right[1]-qp_prev.co[1]) / delta_y), ((qp.handle_left[1]-qp_prev.co[1]) / delta_y)
                return 'LINEAR', 0, 1

            def _new_kp(kp, frame, value):
                kp.co = Vector((frame, value))
                kp.handle_left = Vector((0, 0))
                kp.handle_right = Vector((0, 0))

            _KP = type("_KP", (), {"co": None, "interpolation": "LINEAR",
                                   "handle_left": None, "handle_right": None,
                                   "handle_right_type": "FREE", "handle_left_type": "FREE",
                                   "__init__": lambda self, frame, value: _new_kp(self, frame, value)})

            prev_points = ()
            last_rot = bone.rotation_euler
            key_func = lambda kp: kp.co[0]
            tmp_fcurves = [[], [], []]
            for qw, qx, qy, qz in zip(*[sorted(c.keyframe_points, key=key_func) for c in quaternion_fcurves]):
                rot = Quaternion((qw.co[1], qx.co[1], qy.co[1], qz.co[1])).to_euler(axis_order)
                rot.make_compatible(last_rot)
                frame = qw.co[0]
                curr_points = (qx, qy, qz, qw)
                if prev_points:
                    print("***")
                    interpolation, y1, y2 = _get_bezier_data(prev_points, curr_points)
                    for c, value in zip(tmp_fcurves, rot):
                        ep = _KP(frame, value)
                        c.append(ep)
                        ep.handle_left.x = qw.handle_left.x
                        ep.handle_right.x = qw.handle_right.x
                        ep_prev = c[-2]
                        y_base = ep_prev.co[1]
                        scale = value - y_base
                        ep_prev.interpolation = interpolation
                        ep_prev.handle_right_type = 'FREE'
                        ep_prev.handle_right.y = y1 * scale + y_base
                        ep.handle_left_type = 'FREE'
                        ep.handle_left.y = y2 * scale + y_base
                        print("handle_right x", "cur", ep.handle_right.x, "prev", ep_prev.handle_right.x) 
                else:
                    for c, value in zip(tmp_fcurves, rot):
                        ep = _KP(frame, value)
                        c.append(ep)
                        ep.handle_left.x = qw.handle_left.x
                        ep.handle_right.x = qw.handle_right.x
                prev_points = curr_points
                last_rot = rot
            for c, tc in zip(euler_fcurves, tmp_fcurves):
                for tp in tc:
                    kp = c.keyframe_points.insert(tp.co[0], tp.co[1])
                    kp.interpolation = tp.interpolation
                    kp.handle_left = tp.handle_left
                    kp.handle_right = tp.handle_right
                    kp.handle_left_type = tp.handle_left_type
                    kp.handle_right_type = tp.handle_right_type
```
This works fine.
```
***
handle_right x cur 33.031494140625 prev 0.0
handle_right x cur 33.031494140625 prev 0.0
handle_right x cur 33.031494140625 prev 0.0
***
handle_right x cur 90.6377944946289 prev 33.031494140625
handle_right x cur 90.6377944946289 prev 33.031494140625
handle_right x cur 90.6377944946289 prev 33.031494140625
***
handle_right x cur 185.99212646484375 prev 90.6377944946289
handle_right x cur 185.99212646484375 prev 90.6377944946289
handle_right x cur 185.99212646484375 prev 90.6377944946289
...
```
I am guessing that the internal variable memory of the keyframe will be reallocated when updating the interpolation or handle_type, but the cause could not be determined.

Reply:
Good catch. :+1: 

But I've got an idea to add this support (3e39b0e), it should be easier and better I think. :smile:
(not sure if there is any werid bug though. :worried: )

Reply:
it seems to work fine. thank you :smile: